Source/Server/MqttService.cpp

```diff
@@ -103,57 +84,48 @@ namespace MQTT {
 			
 			clientState->ConnectionFlags = package.VariableHeader.VariableLevel;
 
-			auto shouldDisconnectClient = !(RuleEngine({
-				{new ClientConnectedRule(packageClientId, m_ClientStates), false},
-				{new CorrectProtocolNameRule(protocolName), true},
-				{new Protocol311Rule(package.VariableHeader.Level), true},
-				{new ConnectReservedFlagSetRule(package.VariableHeader.VariableLevel), false},
-				{new IsCredentialFlagIncorrectRule(package.VariableHeader.VariableLevel), false},
-				{new ConnectWillRule(clientState, package.VariableHeader.VariableLevel, package.Payload.WillMessage), true}
-			}).Run());
+			auto action = Protocol::Validators::ConnectValidator()
+				.ValidateClient(package, m_ClientStates, clientState);
+
+			clientState->ConnectionIdentifier = client.GetIdentifier();
 
-			if (shouldDisconnectClient)
+			switch (action)
 			{
+			case MQTT::Protocol::Validators::ConnectValidator::Disconnect:
```

Comment:
should we set client state to disconnected if he is connected already?

Reply:
Yes, this is the action after the validation. So if the package is malformed, we have to disconnect the client from the current session.